examples/ivr_outbound_dialer.py

```diff
@@ -0,0 +1,409 @@
+            log.debug("'{}': Cancel dtmf timeout job".format(sess.uuid))
+            timer.cancel()  # if timer in 'initial' state this is a noop
+        return timer
+
+    def start_dtmf_timer(self, sess, timeout=3):
+        """Start a timer thread which will invoke the dtmf timeout handler after
+        `timeout` seconds. The default interval is 3 seconds.
+        """
+        timer = self.cancel_timer(sess)
+
+        # allocate a new timer
+        timer = threading.Timer(timeout, self.dtmf_timeout_action, [sess])
+        timer.daemon = True  # thread dies with parent process
+        timer.start()  # NOTE: this launches a thread per call
+        sess.vars['dtmf_timer'] = timer
+        return timer
+
+    def dtmf_timeout_action(self, sess):
+        """Timer handler that implements DTMF timeout
+        """
+        call = sess.call
+        log.info("'{}': DTMF timeout".format(sess.uuid))
+
+        if call.vars.get('playing') is True:
+            call.vars['playing'] = False
+            sess.breakmedia()  # stop playback
+
+        # Reset incoming dtmf queue
+        call.vars['incoming_dtmf'] = []
+
+        # Example of playing a prompt urging the user to try again
+        play_filename = '{}/ivr-hello.wav'.format(self.sound_dir)
+        call.vars['playing'] = True
+        sess.playback(play_filename)
+
+        # Trigger dtmf timeout again
+        self.start_dtmf_timer(sess)
+
+    @event_callback('DTMF')
+    def on_digit(self, sess):
+        """Process DTMF digit events
+        """
+        call = sess.call
+        digit = sess['DTMF-Digit']
+        log.info("'{}': DTMF dtmf digit '{}'".format(sess.uuid, digit))
+
+        # Add incoming digit into the digit queue
+        call.vars.setdefault('incoming_dtmf', []).append(digit)
+
+        # DTMF has just been detected, stop playing any files to the user
+        if call.vars.get('playing') is True:
+            sess.breakmedia()
+
+        # Stop the dtmf timeout timer if one exists
+        self.cancel_timer(sess)
+
+        # IVR Menu - a map of digits to actions
+        actions = {
+            '911': '{}/ivr-contact_system_administrator.wav',
+            '811': '{}/ivr-hello.wav',
+            '111': 'hangup',
+        }
+
+        digits = call.vars.get('incoming_dtmf', [])
+        if len(digits) == 3:
+            digits_str = ''.join(digits)
+            log.info("'{}': Matched on dtmf sequence '{}'"
+                     .format(sess.uuid, digits_str))
+            log.info("'{}': Playing file STARTED".format(sess.uuid))
+
+            # lookup IVR action (eg. this can be a call to an external
+            # database / extension list)
+            action = actions.get(digits_str)
+
+            # process action
+            if action is 'hangup':
+                log.info("'{}': User chose to hangup".format(sess.uuid))
+                sess.hangup()
+            elif action:
+                play_filename = action.format(self.sound_dir)
```

Comment:
Totally agree. This was to introduce the idea. I would much prefer to allow the user to define routines which will run on on a particular DID pattern match.